pep/003-dependency-namespaces-and-labels.md

```diff
@@ -0,0 +1,508 @@
+Another example is a meta-bundle composed of stateless micro-services, where when the application is uninstalled dependencies like the logger are more clearly not required anymore and should be uninstalled as well.
+
+### Unmanaged Dependency
+
+Dependencies that are not declared as managed are not upgraded automatically when the current bundle is upgraded.
+This is useful in cases where the dependency is shared with unrelated bundles such as a SQL Server installation.
+
+### Dependency Resolution
+
+🚧
+* Resolve the entire graph, not just 1 level deep
+* Identify shared dependencies in the graph based on match criteria. The same bundle dependency can either be shared between dependencies or some may get their own instance depending on the criteria.
+
+___
+
+## Implementation
+
+<!--
+After the PEP status is changed to implementable, when the PEP has been
+implemented link to the pull request(s) here.
+-->
+
+
+---
+
+## Backwards Compatibility
+
+The dependencies section in porter.yaml is changing its format.
+Currently the list of required bundles is directly underneath dependencies, and this proposal changes it to be under dependencies.requires.
+Since we are pre 1.0, it's an acceptable breaking change but we will need to warn people before the release and call it out in the release notes.
+
+___
+
+## Security Implications
+
+* Reusing outputs from existing installations
+
+  This proposal allows you to reuse outputs from an existing installation.
+  Porter does not have any ACLs on installation data, you either have access to the host environment or you don't.
+  The way to isolate installation data from being accessed or reused by another installation is to use separate host environments.
+  Users could do this previously through `porter installation outputs show OUTPUT --installation INSTALLATION` so it's not a new concern, just makes it easier.
+
+___
+
+## Rejected Ideas
+
+* We considered the seemingly obvious solution of scanning a OCI registry for bundles that implement an interface, and rejected it because:
+  * Use Case: The use case isn't to not have to think about which bundle is used for a dependency but being able to select from a variety of implementations.
+  * Fit for Purpose: The implementation rarely doesn't matter and relies on the target environments infrastructure. Selecting an implementation and bundle author at random wouldn't make anyone happy.
+  * Performance: Scanning registries would significantly slow down bundle installation.
+  * Predictibility: A different bundle could be selected depending on the search order and the contents of the registry on any day.
+
+
+---
+
+## Open Questions
+
+* I'm not sure yet how to handle passing credentials to the dependency that implements an interface, as that isn't possible to standardize in the interface.
```

Comment:
Yea, this is tricky... and ripe for brainstorming!

Reply:
This is now answered in the Q&A section "How can we use an interface when implementations have different credentials?"